embassy-stm32/src/ucpd.rs

```diff
@@ -0,0 +1,581 @@
+    pub fn new(
+        _peri: impl Peripheral<P = T> + 'd,
+        _cc1: impl Peripheral<P = impl Cc1Pin<T>> + 'd,
+        _cc2: impl Peripheral<P = impl Cc2Pin<T>> + 'd,
+    ) -> Self {
+        T::enable_and_reset();
+
+        let r = T::REGS;
+        r.cfgr1().write(|w| {
+            // "The receiver is designed to work in the clock frequency range from 6 to 18 MHz.
+            // However, the optimum performance is ensured in the range from 6 to 12 MHz"
+            // UCPD is driven by HSI16 (16MHz internal oscillator), which we need to divide by 2.
+            w.set_psc_usbpdclk(PscUsbpdclk::DIV2);
+
+            // Prescaler to produce a target half-bit frequency of 600kHz which is required
+            // to produce transmit with a nominal nominal bit rate of 300Kbps+-10% using
+            // biphase mark coding (BMC, aka differential manchester coding).
+            // A divider of 13 gives the target frequency closest to spec (~615kHz, 1.625us)
+            // but we go with the (hopefully well tested) default value used by the Cube HAL
+            // which is 14 divides the clock down to ~571kHz, 1.75us.
+            w.set_hbitclkdiv(14 - 1);
+
+            // Time window for detecting non-idle (12-20us).
+            // 1.75us * 8 = 14us.
+            w.set_transwin(8 - 1);
+
+            // Time from the end of last bit of a Frame until the start of the first bit of the
+            // next Preamble (min 25us).
+            // 1.75us * 17 = ~30us
+            w.set_ifrgap(17 - 1);
+
+            w.set_ucpden(true);
+        });
+
+        Self {
+            cc_phy: CcPhy { _lifetime: PhantomData },
+        }
+    }
+
+    /// Returns the TypeC CC PHY.
+    pub fn cc_phy(&mut self) -> &mut CcPhy<'d, T> {
+        &mut self.cc_phy
+    }
+
+    /// Splits the UCPD driver into a TypeC PHY to control and monitor CC voltage
+    /// and a Power Delivery (PD) PHY with receiver and transmitter.
+    pub fn split_pd_phy(
+        self,
+        rx_dma: impl Peripheral<P = impl RxDma<T>> + 'd,
+        tx_dma: impl Peripheral<P = impl TxDma<T>> + 'd,
+        cc_sel: CcSel,
+    ) -> (CcPhy<'d, T>, PdPhy<'d, T>) {
+        let r = T::REGS;
+
+        // TODO: Currently only SOP messages are supported.
+        r.tx_ordsetr().write(|w| w.set_txordset(0b10001_11000_11000_11000));
+
+        r.cfgr1().modify(|w| {
+            // TODO: Currently only hard reset and SOP messages can be received.
+            w.set_rxordseten(0b1001);
+
+            // Enable DMA
+            w.set_txdmaen(true);
+            w.set_rxdmaen(true);
+        });
+
+        // Enable the receiver on one of the two CC lines.
+        r.cr().modify(|w| {
+            w.set_phyccsel(cc_sel);
+            w.set_phyrxen(true);
+        });
+
+        // Enable hard reset receive interrupt.
+        r.imr().modify(|w| w.set_rxhrstdetie(true));
+
+        // Both parts must be dropped before the peripheral can be disabled.
```

Comment:
wrt refcounting: I think it's possible to reuse what USART already does for splitting: not 100% sure how this works, plus it's been recently refactored, but i thiiink: with [this](https://github.com/embassy-rs/embassy/blob/b2d236ee390081ec6aeef1a27da06098f9febbf9/embassy-stm32/build.rs#L392), `fn enable_and_reset()` and `fn disable()` from `trait RccPeripheral` can do all the refcounting for you, so all you'd need to do in `Drop` impls is `T::disable()`

wrt splitting in general: seems like a good idea. hard to wrap my head around what exactly needs to be seperate from what. naively, CC is just one line, the communication is half duplex, so why would anything need to be seperate at all? but then: it's possible (is it desirable? necessary?) to "preempt" a packet transmit with a hard reset (leading to TXMSGABT? though other parts of the RM imply there is no interrupt for this) or an incoming message (TXMSGDISC. RM mentions "concurrent recieve") that you might not want to miss the content of, so wouldn't those need to be separate futures, maybe running in separate tasks/threads/priorities even? a message being recieved could likewise get interrupted by a hard reset reception, but this seems like it maps pretty cleanly onto returning an error from the rx. (maybe this is why the interrupts are called HRSTSENT not TXHRST and RXHRSTDET not HRSTDET?) and obviously you still need to be detecting a type-c detach while doing PD stuff.

food for thought. probably just my overthinking again >:) I imagine the best way to figure out what exactly is needed is actually implementing the PD negotiations, but that's a fair bit of work

Reply:
Thanks for the refcounting pointers, I think the code you linked is for refcounting of RCC enable its. The "refcounting" bool here is for tracking enable of the peripheral itself.

Preemting a transmit package is possible with the current API:
1. Drop the ongoing transmit future (it will continue to run in the background, with trash data because the DMA stopped).
2. Send a hard reset: The peripheral interrupts the still running transmission

Regarding splitting you made me overthink this *again* and now I think it might be feasible to implement all of the PD spec without any split.
Pro: No refcounting for drop needed.
Con: Runtime selection of PD comminaciton pin required before TX or RX can be used -> Introduces a new error code. This error condition is currently nicely prevented by the typesystem.

>I imagine the best way to figure out what exactly is needed is actually implementing the PD negotiations, but that's a fair bit of work
Agree!

Reply:
> Thanks for the refcounting pointers, I think the code you linked is for refcounting of RCC enable its. The "refcounting" bool here is for tracking enable of the peripheral itself.

hm. Thought these were the same, but now I’ve done even more thinking I’m less sure. I think the reason @Dirbaio generalised the ref counting is because some bits in the RCC registers are shared between several peripherals. Not 100% sure how “disabled” a peripheral is without its clock enabled, but my guess would be  “not fully”. Wonder if this means USART is wrong then, because the UE bit isn’t actually touched anywhere in any Drop impls? UCPD has more asynchronous stuff going on as well e.g. pullups 
If all the “functionality” is disabled, by e.g. CCENABLE and stuff when dropping, or not transmitting on your UART, which you can’t be while dropping it, is it fine? when creating a new one everything should be fixed by the reset, but what about between the clock enable and the reset?

Reply:
Ohh, I misunderstood how the RCC ref-counting works. In any case the imlementation implements at least on call to `T::disable()`.
Let me try to disable as much as possible without internal ref-counting and then hope that turning of the clock shuts down the rest (even when not clearing UCPDEN).